src/start/AuthScreen.js

```diff
@@ -220,12 +231,47 @@ class AuthScreen extends PureComponent<Props> {
     this.props.dispatch(navigateToPassword(serverSettings.require_email_format_usernames));
   };
 
-  handleAuth = (method: AuthenticationMethodDetails) => {
+  handleNativeAppleAuth = async () => {
+    const { dispatch, realm } = this.props;
+    const nativeAppleOTP = await webAuth.generateOtp();
+    const credential = await AppleAuthentication.signInAsync({
+      state: nativeAppleOTP,
+    });
+    if (credential.state !== nativeAppleOTP) {
+      throw new Error('OTP mismatch');
+    }
+
+    // TODO: handle errors from this fetch (no Zulip account exists, etc.)
+    const { url: callbackUrl } = await fetch(
+      `${this.props.realm}/complete/apple/?mobile_flow_otp=${nativeAppleOTP}`,
+      {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+        },
+        body: encodeParamsForUrl(credential),
+      },
+    );
+
```

Comment:
I don't think this will work as is. I'm not sure we want to be using the `/complete/apple/` route at all.

Here's what Apple will cause to be sent to that route (under "Handle the Response"):
https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_js/incorporating_sign_in_with_apple_into_other_platforms

Here's the data we have at this point, in the `AppleAuthenticationCredential` (from Expo, and corresponding closely to [this](https://developer.apple.com/documentation/authenticationservices/asauthorizationappleidcredential) from Apple):
https://docs.expo.io/versions/latest/sdk/apple-authentication/#appleauthenticationappleauthenticationcredential

Hmm, looking closer, I guess they actually align quite well in what information is present. They just aren't quite the same shape -- so that line `body: encodeParamsForUrl(credential)` will need to be expanded to do a bit more work :wink: 

Then, one thing that simplifies things (and may also call for adding a separate route) is: we don't have a need for the redirect to `zulip://`, or the OTP mechanism which exists to encrypt the API key through that channel. That redirect is needed for getting from a browser back to the app's code. If we're making our own `fetch` here, we can just get the data from that directly.

Alternatively, for the new-account case maybe we do want to open a browser. If we do open a browser, then we need the `zulip://` redirect and the `mobile_flow_otp` mechanism. For simplicity we might, at least at first, just always open a browser, regardless of whether the user might already have an account. In that case we'll want to call something like `webAuth.openBrowser` instead of `fetch`.

Reply:
> For simplicity we might, at least at first, just always open a browser, regardless of whether the user might already have an account. In that case we'll want to call something like `webAuth.openBrowser` instead of `fetch`.

Mmm, this does make sense as a place to start; I was [a bit worried](https://github.com/zulip/zulip-mobile/pull/4034#issuecomment-612268425) about how to handle basically any case other than the completely successful login path, with the browserless `fetch` way. All those contingencies (no Zulip account, etc.) could be handled in the browser. We would indeed need the `zulip://` redirect, but I'm not sure I've understood whether you've decided to add the separate route and use it here.

Using the browser for for everyone is not ideal for the long-term, though, right; users with an existing account would want to log in seamlessly, without the browser. I'd basically want to open an issue for this immediately upon merging this simpler approach; does that sound right? (And I hope opening the browser for people with existing accounts won't be enough to trigger an App Store review rejection.) One way this more complicated, but smoother, process could look is

1. We use `fetch` on either a) a route that never gives the `zulip://` redirect, since, as you've said, we don't need it (in this case, adding a route is necessary), or b) a route that we can pass a flag to, saying we don't want the redirect (here, it would be possible, but maybe awkward, to use an altered form of the existing route).
2. Then, the result would either contain an API key, and we'd be ready to go, or (e.g., if there's no Zulip account) a signal that we need to try again, this time with `webAuth.openBrowser`.

Reply:
> Using the browser for everyone is not ideal for the long-term, though, right; users with an existing account would want to log in seamlessly, without the browser. I'd basically want to open an issue for this immediately upon merging this simpler approach; does that sound right?

Yeah, that sounds right.


> (And I hope opening the browser for people with existing accounts won't be enough to trigger an App Store review rejection.)

Yeah, me too. Good thing we're not right up against the deadline.


> One way this more complicated, but smoother, process could look is [...]

This sounds like a good plan. The `mobile_flow_otp` option already serves as a flag that controls whether it does the `zulip://` redirect, vs. redirecting to the webapp (in addition to providing a one-time pad to use in the `zulip://` redirect.) So I think it'd probably make sense to go for (b) and just add another flag with a meaning like "just return the API key in a JSON response". But @mateuszmandera may have more to say about what'd be a convenient arrangement on the server side.

Reply:
Hmm, so the plan would be to do the browser flow only for now?

For the native approach - I think `/complete/apple/` is still a good endpoint to use for this as on the server side we can override the class method handling the endpoint to be able to deal with the "native flow" request as well. The upside of that being that then the rest of the pipeline for processing "social authentication" can run as normal (avoiding code duplication). And then we'd have two cases for the response you'll get from the server:
1. Complete success - you get the `zulip://` redirect (yes, it's kind of redundant to do the OTP thing, but on the other hand - code reuse is nice)
2. Some kind of failure - you get a https redirect. If you get this, you could open the redirect in the browser and that will handle the error. "Bad" errors will get the same handling as in browser flow (show an error message if we're currently doing that in the browser flow). And the "user doesn't have an account" will take the user through the registration flow in the browser and upon success redirect back to the app with the newly created account through the `zulip://` redirect.

How does the above sound? @gnprice  @chrisbobbe

Reply:
> Hmm, so the plan would be to do the browser flow only for now?

Yes but there's a couple of things that can mean, so to be explicit:
* For users on Android and on iOS <=12, we'll do the whole auth flow in a browser, same way we do Google and GitHub and SAML and the rest.
* For users on iOS >=13 where we can, we'll
  * use the system-provided flow to get a credential from the user -- so the user gets some system UI that says like "do you want to sign into this app with your identity, what name do you want, do you want to use your real email address or an opaque alias", and when they say yes we get one of these [`AppleAuthenticationCredential`](https://docs.expo.io/versions/latest/sdk/apple-authentication/#appleauthenticationappleauthenticationcredential) objects;
  * then we'll turn that into a URL like `https://example.zulipchat.com/complete/apple?id_token=…&mobile_flow_otp=…&…` and open a browser with that URL, and pick up the browser flow from there. In particular if the user already has an account and the server accepts the login, it'll redirect the browser to a `zulip://…` URL with the encrypted API key.


> Some kind of failure - you get a https redirect. If you get this, you could open the redirect in the browser and that will handle the error. "Bad" errors will get the same handling as in browser flow (show an error message if we're currently doing that in the browser flow).

Hmm. A couple of questions/thoughts I'd have are:
* Will the URL in the redirect be enough to just pass to a fresh browser and it'll pick up the flow smoothly from there? No cookies or other state needed?
* That seems like it wouldn't cleanly support the app trying to understand anything about the failure and give a structured error of its own -- it'd just always have to go to a browser. That might be OK, but it puts a ceiling on how polished the experience can be with this approach.

In any case I think we can discuss more and experiment when we get there. I would really like to get to a point where we have one flow that works solidly; among other things I think that will help us a lot in concretely understanding how all the steps of this are supposed to work. So I will be very happy for us to just get the flow with `/complete/apple/` in a browser working.